### DISCUSSION TITLE
Windows Community Toolkit 8.0.230823-rc Feedback

### DISCUSSION THREAD

**Discussion by michael-hawker** [CommunityToolkit/Windows#184] [Announcements]
Today, we're happy to ship a pre-release version of the Windows Community Toolkit, version 8.0.230823-rc. 🎉🎉🎉

_You can find out more about this release and how to use it on our blog post._ (Link in a few minutes)

You can find these packages of the new version of the Toolkit on NuGet.org for the very first time! Just be sure to check "**Include Prerelease**" next to the NuGet Package Manager search box. The available packages are:

- **Animations** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Animations) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Animations)
- **Behaviors** ✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Behaviors) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Behaviors) ✨
- **Collections** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Collections) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Collections)
- Controls - All the following packages are under the `CommunityToolkit.WinUI.Controls` namespace:
  - **CameraPreview** 💻 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.CameraPreview) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.CameraPreview)
  - **HeaderedControls** ✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.HeaderedControls) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.HeaderedControls) ✨
  - **ImageCropper** 💻 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.ImageCropper) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.ImageCropper)
  - **LayoutTransformControl** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.LayoutTransformControl) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.LayoutTransformControl)
  - **MetadataControl** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.MetadataControl) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.MetadataControl)
  - **Primitives** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.Primitives) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.Primitives)
  - **RadialGauge** 💻🎨✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.RadialGauge) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.RadialGauge) ✨
  - **RangeSelector** 🎨 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.RangeSelector) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.RangeSelector)
  - **RichSuggestBox** 💻🎨 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.RichSuggestBox) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.RichSuggestBox)
  - **Segmented** ✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.Segmented) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.Segmented) ✨
  - **SettingsControls** ✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.SettingsControls) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.SettingsControls) ✨
  - **Sizers** 🎨✨ - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.Sizers) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.Sizers) ✨
  - **TokenizingTextBox** 🎨 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Controls.TokenizingTextBox) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Controls.TokenizingTextBox)
- **Converters** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Converters) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Converters)
- **DeveloperTools** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.DeveloperTools) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.DeveloperTools)
- **Extensions** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Extensions) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Extensions)
- **Helpers** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Helpers) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Helpers)
- **Lottie** 💻 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Lottie) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Lottie)
- **Media** 💻 - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Media) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Media)
- **Triggers** - [WinUI 2/UWP](https://nuget.org/packages/CommunityToolkit.Uwp.Triggers) - [WinUI 3/WindowsAppSDK](https://nuget.org/packages/CommunityToolkit.WinUI.Triggers)

    '✨' denotes new or updated packages/features from version 7.1.
    '🎨' denotes updated XAML styling from version 7.1.
    '💻' denotes packages which only target _UWP/WindowsAppSDK_. However, its absence does not imply that every feature will work across all platforms (see note in introduction).

🔥🐉 **Note**: _Mixing and matching_ these new 8.0 packages with existing Windows Community Toolkit 7.x packages, especially UI controls, can lead to issues with the namespace changes and dependencies on common code (like the Visual Tree extensions). If there's something that we haven't yet ported but you need, [open a discussion](https://github.com/CommunityToolkit/Windows/discussions) to let us know or claim it and contribute!

Please use this discussion for any feedback or issues you encounter with this release.

We'll be working to update our docs, website, and sample app over the coming week or so before a final release.

**Full Changelog**: https://github.com/CommunityToolkit/Windows/commits/v8.0.230823-rc

<hr /><em>This discussion was created from the release <a href='https://github.com/CommunityToolkit/Windows/releases/tag/v8.0.230823-rc'>v8.0.230823-rc</a>.</em>

**Comment by kmgallahan** [comment]
Can the Labs Sample App be built using this feed? Last time I tried there were build errors using the special feed.

**Reply by michael-hawker** [reply]
We have to update Labs to use the latest NuGet drop still, is that what you mean?

**Reply by kmgallahan** [reply]
2-3 weeks ago I downloaded the latest Labs Sample App code, tried to point it to the labs feed since it was referencing those packages, but still ran into a pile of build errors.

**Reply by michael-hawker** [reply]
Once we ship this release, we'll get back to ensuring Labs is back up-to-date and in-sync.

**Comment by kmgallahan** [comment]
Upgrading to anything past `8.0.230811-preview` results in many of these errors for me:

```
The type 'WindowEventArgs' exists in both 'Microsoft.WinUI, Version=3.0.0.0, Culture=neutral, PublicKeyToken=de31ebe4ad15742b' and 'Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null'
```

Don't know why Uno is being referenced in my WinUI projects.

Edit:

Also see this in build output for a WinUI controls lib that references the toolkit:

```
(Uno.UI.SourceGenerators.XamlGenerator.XamlParsingException: An error was found in Style ---> System.InvalidOperationException: Unable to find type Microsoft.UI.Xaml.Controls.TextBlock
```

And after lots of trying I still can't build with many of these errors:

```
XamlCompiler error WMC0901: A type with a full name of 'Microsoft.UI.Xaml.FrameworkElement' already exists in an included reference.
```

**Reply by Martin1994** [reply]
Is this because `net7.0` is preferred over `net6.0-windows10.0.19041` when the target framework is `net8.0-windows10.0.19041`?

**Reply by michael-hawker** [reply]
`8.0.230828-rc` packages have been pushed to NuGet which should resolve this issue. Thanks for reporting!

**Comment by Mysterious-Dev** [comment]
That's weird, when I add this package, I have this 2 errors on app startup (my app is generated with template studio 5.4) :/

![image](https://github.com/CommunityToolkit/Windows/assets/40738104/a9fb33ce-9f56-4409-966d-43e84f7be8c2)

**Reply by michael-hawker** [reply]
Thanks for providing feedback, this looks a bit different (though still similar) to the other errors reported here. Can you provide a bit more details about which project is failing and what it's TFM/configuration is?

**Reply by michael-hawker** [reply]
Can you try the latest `8.0.230828-rc` package we just pushed to NuGet and see if you see the same issue or not? If you still have an issue, please open an issue on the repo with more details ASAP. Thanks!

**Reply by Lightczx** [reply]
It's fine now to have these packages
```xml
    <PackageReference Include="CommunityToolkit.Labs.WinUI.Controls.DataTable" Version="0.1.230802" />
    <PackageReference Include="CommunityToolkit.Labs.WinUI.Shimmer" Version="0.1.230802" />
    <PackageReference Include="CommunityToolkit.Labs.WinUI.TokenView" Version="0.1.230802" />
    <PackageReference Include="CommunityToolkit.Labs.WinUI.TransitionHelper" Version="0.1.230802" />
    <PackageReference Include="CommunityToolkit.Mvvm" Version="8.2.2-build.1" />
    <PackageReference Include="CommunityToolkit.WinUI.Behaviors" Version="8.0.230828-rc" />
    <PackageReference Include="CommunityToolkit.WinUI.Controls.HeaderedControls" Version="8.0.230828-rc" />
    <PackageReference Include="CommunityToolkit.WinUI.Controls.Primitives" Version="8.0.230828-rc" />
    <PackageReference Include="CommunityToolkit.WinUI.Controls.Segmented" Version="8.0.230828-rc" />
    <PackageReference Include="CommunityToolkit.WinUI.Controls.SettingsControls" Version="8.0.230828-rc" />
    <PackageReference Include="CommunityToolkit.WinUI.Media" Version="8.0.230828-rc" />
    <PackageReference Include="CommunityToolkit.WinUI.Notifications" Version="7.1.2" />
    <PackageReference Include="CommunityToolkit.WinUI.UI.Controls.Core" Version="7.1.2" />
```
And I'm wonder when the `Core` package will get ported to the new namespace,
Any plan? @michael-hawker 
Thanks in advance.

**Reply by Mysterious-Dev** [reply]
@michael-hawker The latest version solves the problem. Thanks ^^.

**Reply by michael-hawker** [reply]
> And I'm wonder when the Core package will get ported to the new namespace,

@Mysterious-Dev there's not a 1:1 mapping between packages from 7.x to 8.0, probably should call that out more clearly in our release blog or docs.

8.0 is more finely grained so things like `CameraPreview` and `MetadataControl` have their own packages. `RadialProgressBar` is deprecated as it's in WinUI's options for `ProgressRing` now, similar to `InAppNotification` we have a `StackedNotificationsBehavior` in the Behaviors package on top of `InfoBar`. `DropShadowPanel` was replaced with the Attached Shadows in the `Extensions` package.

**Reply by Lightczx** [reply]
I really wish to have ImageEx ported, my project heavily depending on the LazyLoading and init Animation the control bring in.

**Reply by michael-hawker** [reply]
@Lightczx feel free to open an issue to track and garner support or help work with us to bring it over from the old repository. It just wasn't high enough on our priority list of top-used components to make the cut for our initial work. There's been a lot added to the Toolkit over the years!

It's also an older component in the Toolkit and could be due for some possible re-architecturing. We've had folks try to use it in mass galleries which wasn't its original intent and other folks which want custom caching strategies, etc...

**Comment by kmgallahan** [comment]
I would recommend that the Primitives package get a more descriptive NuGet Readme. 

> This package contains Primitives.

Finding where DockPanel lived was a case of: "well... it isn't any of these other things, so I guess I'll check the Primitives source code to figure out what's in it".

**Reply by michael-hawker** [reply]
Yeah, I realized we just had generic descriptions when I first pushed to NuGet. I opened #185 so we can add more here for each package.

**Comment by zhuxb711** [comment]
Hope BladeView and Markdown control will be included in CommunityToolkit.WinUI.X 8.0

**Reply by michael-hawker** [reply]
BladeView is pretty far down on the list of widely used controls. It's a fairly uncommon pattern. So, it'd be something to open a separate discussion about, especially as it'd need a visual refresh as well. (Similar to [this comment](https://github.com/CommunityToolkit/Windows/discussions/184#discussioncomment-6860035)).

For Markdown, it's not included in the 8.0 release, it's a very old component which was using a custom Markdown parser. The good news is that some folks have been working on a [replacement in Windows Community Toolkit Labs](https://github.com/CommunityToolkit/Labs-Windows/pull/480) that uses Markdig! It just needs a bit more work to fix some issue with the CI once we get the Labs-Windows repo back up-to-date with our latest tooling changes. Our focus the last few weeks has been on this release, so we'll get back to Labs-Windows more after and try and get that one moving along again.

**Reply by zhuxb711** [reply]
> BladeView is pretty far down on the list of widely used controls. It's a fairly uncommon pattern. So, it'd be something to open a separate discussion about, especially as it'd need a visual refresh as well. (Similar to [this comment](https://github.com/CommunityToolkit/Windows/discussions/184#discussioncomment-6860035)).
> 
> 
> 
> For Markdown, it's not included in the 8.0 release, it's a very old component which was using a custom Markdown parser. The good news is that some folks have been working on a [replacement in Windows Community Toolkit Labs](https://github.com/CommunityToolkit/Labs-Windows/pull/480) that uses Markdig! It just needs a bit more work to fix some issue with the CI once we get the Labs-Windows repo back up-to-date with our latest tooling changes. Our focus the last few weeks has been on this release, so we'll get back to Labs-Windows more after and try and get that one moving along again.

Thanks for quick response. BladeViewer is very useful if someone wants to create multiple panels just like Edge, so it's not a uncommon pattern and should be reconsidered.
